wp-includes/rest-api/endpoints/class-wp-rest-customize-controls-controller.php

```diff
@@ -0,0 +1,371 @@
+	}
+
+	/**
+	 * Checks if a given request has access to read controls.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {
+		return current_user_can( 'edit_theme_options' );
+	}
+
+	/**
+	 * Retrieves list of controls.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$wp_customize = $this->ensure_customize_manager();
+
+		$all_controls = $wp_customize->controls();
+		$controls = array();
+
+		foreach ( $all_controls as $control_id => $control ) {
+			if ( ! $control->check_capabilities() ) {
+				continue;
+			}
+			$data = $this->prepare_item_for_response( $control, $request );
+			$controls[] = $this->prepare_response_for_collection( $data );
+		}
+
+		return rest_ensure_response( $controls );
+	}
+
+	/**
+	 * Prepares a single control's output for response.
+	 *
+	 * @since 4.?.?
+	 * @access public
+	 *
+	 * @param WP_Customize_Control $control WP_Customize_Control object.
+	 * @param WP_REST_Request      $request Request object.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function prepare_item_for_response( $control, $request ) {
+
+		$control_array = (array) $control;
+		$data = array();
+		$primary_setting = '';
+
+		$links = array(
+			'self' => array(
+				'href' => rest_url( trailingslashit( $this->namespace . '/' . $this->rest_base ) . $control_array['id'] ),
+			),
+		);
+
+		$hide_from_response = array(
+			'capability',
+			'instance_number',
+			'manager',
+			'json',
+			'setting', // Hide this from response since that's returned within 'settings'.
+			'active_callback',
+		);
+
+		// Remove unused params of upload control.
+		if ( $control instanceof WP_Customize_Upload_Control ) {
+			$hide_from_response[] = 'removed';
+			$hide_from_response[] = 'context';
+			$hide_from_response[] = 'extensions';
+		}
+
+		if ( ! empty( $control_array['section'] ) ) {
+			$links['up'] = array(
+				'href' => rest_url( trailingslashit( $this->namespace ) . 'sections/' . $control_array['section'] ),
+				'embeddable' => true,
+			);
+		}
+
+		// Get primary setting ID.
+		if ( is_object( $control_array['setting'] ) ) {
+			$primary_setting = $control_array['setting']->id;
+			$data['settings'] = array( $primary_setting );
+			$links['related'] = array( array(
+					'href' => rest_url( trailingslashit( $this->namespace ) . 'settings/' . $primary_setting ),
+					'embeddable' => true,
+				),
+			);
+		}
+
+		foreach ( $control_array as $property => $value ) {
+			if ( in_array( $property, $hide_from_response, true ) ) {
+				continue;
+			} elseif ( '' === $value ) {
```

Comment:
@miina humm. I think setting all props that are empty strings to `null` will not have the expected result. I think you should probably list out the props by name specifically for the ones that we want to nullify, similar to what you did with `$hide_from_response`.